File: JS101_programming_foundations/small_problems/06_easy_6/ex_08.js
// Create a function that takes two integers as arguments.The first argument is a count, and the second is the starting number of a sequence that your function will create.The function should return an array containing the same number of elements as the count argument.The value of each element should be a multiple of the starting number.

// You may assume that the count argument will always be an integer greater than or equal to 0. The starting number can be any integer.If the count is 0, the function should return an empty array.



function sequence(num, start) {
  let array = [];
  for (let i = 0; i < num; i += 1) {
    array.push(start);
    start += start;
  }
  return array;
}

function sequence(num, start) {
  return Array(num).fill().map((_, idx) => start * (idx + 1));
}

sequence(5, 1);          // [1, 2, 3, 4, 5]
sequence(4, -7);         // [-7, -14, -21, -28]
sequence(3, 0);          // [0, 0, 0]
sequence(0, 1000000);    // []

File: RB130/exercises/easy_2/ex_08.rb
def max_by(array)
  return nil if array.empty?

  max_num = array.first
  array.each do |num|
    max_num = num if yield(num) > yield(max_num)
  end

  max_num
end

# def max_by(array)
#   return nil if array.empty?

#   max_num = array.first
#   largest = yield(max_num)

#   array[1..-1].each do |num|
#     if yield(num) > largest
#       largest = yield(num)
#       max_num = num
#     end
#   end

#   max_num
# end

puts max_by([1, 5, 3]) { |value| value + 2 } == 5
puts max_by([1, 5, 3]) { |value| 9 - value } == 1
max_by([1, 5, 3]) { |value| (96 - value).chr } == 1
puts max_by([[1, 2], [3, 4, 5], [6]]) { |value| value.size } == [3, 4, 5]
puts max_by([-7]) { |value| value * 3 } == -7
puts max_by([]) { |value| value + 5 } == nil
